tests/csapi/apidoc_room_event_test.go

```diff
@@ -0,0 +1,206 @@
+	var todoRooms = make(chan bool, goConcurrency)
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				_, ok := <-todoRooms
+
+				if !ok {
+					return
+				}
+
+				roomId := alice.CreateRoom(t, map[string]interface{}{
+					"preset": "public_chat",
+				})
+
+				doneRooms <- roomId
+			}
+		}()
+	}
+
+	for i := 0; i < roomAmount; i++ {
+		todoRooms <- true
+	}
+
+	close(todoRooms)
+
+	var rooms []string
+
+	for i := 0; i < roomAmount; i++ {
+		select {
+		case newRoom := <-doneRooms:
+			rooms = append(rooms, newRoom)
+		case <-time.After(channelTimeout):
+			t.Fatalf("Room creation timed out, got to %d rooms (from %d)", i, roomAmount)
+		}
+	}
+
+	// Send events to all rooms with corresponding roomID
+
+	// get current next_batch
+	_, since := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+	type txnAndId struct {
+		txn    int
+		roomId string
+	}
+
+	var doneEvents = make(chan string, roomAmount)
+	var todoEvents = make(chan txnAndId, goConcurrency)
+
+	txnCount := 0
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				tid, ok := <-todoEvents
+
+				if !ok {
+					return
+				}
+
+				paths := []string{"_matrix", "client", "r0", "rooms", tid.roomId, "send", "m.room.message", strconv.Itoa(tid.txn)}
+				res := alice.MustDoFunc(t, "PUT", paths, client.WithJSONBody(t, map[string]interface{}{
+					"msgtype": "m.text",
+					"body":    tid.roomId,
+				}))
+				body := client.ParseJSON(t, res)
+				eventID := client.GetJSONFieldStr(t, body, "event_id")
+
+				doneEvents <- eventID
+			}
+		}()
+	}
+
+	for _, id := range rooms {
+		todoEvents <- txnAndId{
+			roomId: id,
+			txn:    txnCount,
+		}
+
+		txnCount++
+	}
+
+	close(todoEvents)
+
+	for i := 0; i < roomAmount; i++ {
+		select {
+		case <-doneEvents:
+			break
+		case <-time.After(channelTimeout):
+			t.Fatalf("Receiving events timed out, got to %d events (from %d)", i, roomAmount)
+		}
+	}
+
+	// Collect all events, check if room IDs check out
+
+	okayRooms := make(map[string]bool)
+
+	alice.MustSyncUntil(t, client.SyncReq{Since: since}, func(_ string, jsonObj gjson.Result) error {
+		join := jsonObj.Get("rooms.join")
+
+		if !join.Exists() {
+			return fmt.Errorf("could not find rooms.join")
+		}
+
+		if !join.IsObject() {
+			t.Fatalf("expected join in sync to be object, was %s; %s", join.Type.String(), join)
+		}
+
+		for roomID, room := range join.Map() {
+			if !stringSliceContains(rooms, roomID) {
+				// We're running an isolated test, so we shouldn't be getting sync from other rooms
+				t.Fatalf("Got sync for unrelated room %s; %s", roomID, room)
+			}
+
+			eventsJson := room.Get("timeline.events")
+
+			if !eventsJson.Exists() {
+				continue
+			}
+
+			if !eventsJson.IsArray() {
+				t.Fatalf("Expected events array in sync response to be an array: %s", eventsJson)
+			}
+
+			events := eventsJson.Array()
+
+			if len(events) > 1 {
+				t.Fatalf("Expected 0 or 1 events, got %d events: %s", len(events), events)
```

Comment:
Is this fair to say? We don't ever set a timeline_limit filter. These rooms are freshly made, so will include the initial set of state events, so it could entirely reasonably be >1 events.

EDIT: Oh but we are using a since token that is after the initial room creations, right. Needs a comment!

Reply:
This bit is also the exact invariant the original sytest test was testing; make sure events go down the correct room.

I'm fairly sure that this test will not fail on modern servers anyways, but I also have no compelling reason to remove it, as programmer mistakes can still exist, and this is essentially a canary for other problems.